diesel/src/query_builder/upsert/conditionally_insert_statement.rs

```diff
@@ -0,0 +1,108 @@
+use std::fmt::Debug;
+
+use crate::backend::{Backend, DieselReserveSpecialization};
+use crate::query_builder::upsert::on_conflict_clause::OnConflictValues;
+use crate::query_builder::where_clause::{NoWhereClause, WhereAnd};
+use crate::query_builder::{
+    where_clause, AsQuery, AstPass, InsertStatement, QueryFragment, QueryId,
+};
+use crate::query_dsl::filter_dsl::FilterDsl;
+use crate::sql_types::BoolOrNullableBool;
+use crate::{Expression, QueryResult, QuerySource, RunQueryDsl};
+
+#[derive(Debug, Clone)]
+pub struct ConditionallyInsertStatement<Stmt, WhereClause = NoWhereClause> {
+    insert: Stmt,
+    where_clause: WhereClause,
+}
+
+impl<Stmt, WhereClause> ConditionallyInsertStatement<Stmt, WhereClause> {
+    pub fn new(
+        insert: Stmt,
+        where_clause: WhereClause,
+    ) -> ConditionallyInsertStatement<Stmt, WhereClause> {
+        ConditionallyInsertStatement {
+            insert,
+            where_clause,
+        }
+    }
+}
+
+impl<Stmt, WhereClause> QueryId for ConditionallyInsertStatement<Stmt, WhereClause>
```

Comment:
We should likely just add a `#[derive(QueryId)]` to the struct itself, as the `QueryId` implementation should depend on both struct fields.